packages/iov-ethereum/src/ethereumconnection.ts

```diff
@@ -119,18 +120,24 @@ export class EthereumConnection implements AtomicSwapConnection {
   }
 
   private readonly pollIntervalMs: number;
-  private readonly rpcClient: HttpJsonRpcClient;
+  private readonly rpcClient: JsonRpcClient;
   private readonly myChainId: ChainId;
-  private readonly socket: StreamingSocket | undefined;
   private readonly scraperApiUrl: string | undefined;
   private readonly atomicSwapEtherContractAddress?: Address;
   private readonly erc20Tokens: ReadonlyMap<TokenTicker, Erc20Options>;
   private readonly erc20ContractReaders: ReadonlyMap<TokenTicker, Erc20Reader>;
   private readonly codec: EthereumCodec;
 
   public constructor(baseUrl: string, chainId: ChainId, options: EthereumConnectionOptions) {
+    const baseUrlIsHttp = ["http://", "https://"].some(prefix => baseUrl.startsWith(prefix));
+    const baseUrlIsWs = ["ws://", "wss://"].some(prefix => baseUrl.startsWith(prefix));
+    const wsUrl = baseUrlIsWs ? baseUrl : options.wsUrl;
+    if (!wsUrl && !baseUrlIsHttp) {
+      throw new Error("Unsupported protocol for baseUrl: must be one of http, https, ws or wss");
```

Comment:
I think this error messge should be based on `baseUrlIsWs`/`baseUrlIsHttp` only, so `wsUrl` can be set below this block.